sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsCheckpointMark.java

```diff
@@ -34,6 +34,7 @@
 public class JmsCheckpointMark implements UnboundedSource.CheckpointMark {
 
   private final List<Message> messages = new ArrayList<>();
+  private long oldestPendingTimestamp = System.currentTimeMillis();
```

Comment:
* Suggest storing this as a Joda-time `Instant`.
* Suggest using [`BoundedWindow.TIMESTAMP_MIN_VALUE`](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/BoundedWindow.java#L38) as this is earlier than all timestamps. If you boot up and read old messages (before `now`), you don't want to mark the first few messages as LATE!